source/extensions/transport_sockets/alts/tsi_socket.cc

```diff
@@ -259,28 +263,84 @@ Network::IoResult TsiSocket::doRead(Buffer::Instance& buffer) {
   return repeatReadAndUnprotect(buffer, result);
 }
 
+Network::IoResult TsiSocket::repeatProtectAndWrite(Buffer::Instance& buffer, bool end_stream) {
+  uint64_t total_bytes_written = 0;
+  Network::IoResult result = {Network::PostIoAction::KeepOpen, 0, false};
+
+  ASSERT(prev_handshake_bytes_to_drain_ == 0);
+  while (true) {
+    uint64_t bytes_to_drain_this_iteration =
+        prev_bytes_to_drain_ > 0
+            ? prev_bytes_to_drain_
+            : std::min(buffer.length(), actual_frame_size_to_use_ - frame_overhead_size_);
+    // Consumed all data. Exit.
+    if (bytes_to_drain_this_iteration == 0) {
+      break;
+    }
+    // Short write did not occur previously.
+    if (raw_write_buffer_.length() == 0) {
+      ASSERT(frame_protector_);
+      ASSERT(prev_bytes_to_drain_ == 0);
+
+      // Do protect.
+      ENVOY_CONN_LOG(debug, "TSI: protecting buffer size: {}", callbacks_->connection(),
+                     bytes_to_drain_this_iteration);
+      tsi_result status = frame_protector_->protect(
+          grpc_slice_from_static_buffer(buffer.linearize(bytes_to_drain_this_iteration),
+                                        bytes_to_drain_this_iteration),
+          raw_write_buffer_);
+      ENVOY_CONN_LOG(debug, "TSI: protected buffer left: {} result: {}", callbacks_->connection(),
+                     bytes_to_drain_this_iteration, tsi_result_to_string(status));
+    }
+
+    // Write raw_write_buffer_ to network.
+    ENVOY_CONN_LOG(debug, "TSI: raw_write length {} end_stream {}", callbacks_->connection(),
+                   raw_write_buffer_.length(), end_stream);
+    result = raw_buffer_socket_->doWrite(raw_write_buffer_, end_stream && (buffer.length() == 0));
+
+    // Short write. Exit.
+    if (raw_write_buffer_.length() > 0) {
+      prev_bytes_to_drain_ = bytes_to_drain_this_iteration;
+      break;
+    } else {
+      buffer.drain(bytes_to_drain_this_iteration);
+      prev_bytes_to_drain_ = 0;
+      total_bytes_written += bytes_to_drain_this_iteration;
+    }
+  }
+
+  return {result.action_, total_bytes_written, false};
+}
+
 Network::IoResult TsiSocket::doWrite(Buffer::Instance& buffer, bool end_stream) {
+  Network::IoResult result = {Network::PostIoAction::KeepOpen, 0, false};
   if (!handshake_complete_) {
     Network::PostIoAction action = doHandshake();
     ASSERT(action == Network::PostIoAction::KeepOpen);
     // TODO(lizan): Handle synchronous handshake when TsiHandshaker supports it.
-  }
-
-  if (handshake_complete_) {
+    if (raw_write_buffer_.length() > 0) {
+      ENVOY_CONN_LOG(debug, "TSI: raw_write length {} end_stream {}", callbacks_->connection(),
+                     raw_write_buffer_.length(), end_stream);
+      result = raw_buffer_socket_->doWrite(raw_write_buffer_, end_stream && (buffer.length() == 0));
```

Comment:
The coverage report shows that this branch is not covered by tests.

Reply:
Ah nice finding! It turns out this code path should not be triggered at all. Because in an async TSI, the buffer to be sent to peer (i.e., raw_write_buffer_) should be filled by the handshake service and sent to the peer in the callback function (i.e., doHandshakeNextDone) upon finishing the process of handshake request. After scheduling a handshake request, we do not know if raw_write_buffer_ is ready to be sent to its peer so should not attempt the network write.